include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -157,32 +161,291 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+						// Can't static cast here?
+						hash = nbl::core::XXHash_256((uint8_t*)(hashable.data()), hashable.size() * (sizeof(char) / sizeof(uint8_t)));
+					}
+
+					SPreprocessingDependency(SPreprocessingDependency&) = delete;
+					SPreprocessingDependency& operator=(SPreprocessingDependency&) = delete;
+				
+					inline SPreprocessingDependency(SPreprocessingDependency&&) = default;
+					inline SPreprocessingDependency& operator=(SPreprocessingDependency&&) = default;
+
+					inline bool operator==(const SPreprocessingDependency& other) const
+					{
+						return hash == other.hash && identifier == identifier && contents == contents;
+					}
+
+					// Needed for json vector serialization I believe
+					SPreprocessingDependency() {}
+
+					// path or identifier
+					system::path requestingSourceDir = "";
+					std::string identifier = "";
+					// file contents
+					std::string contents = "";
+					// hash of the contents
+					std::array<uint64_t, 4> hash = {};
+					// If true, then `getIncludeStandard` was used to find, otherwise `getIncludeRelative`
+					bool standardInclude = false;
+					nbl::system::IFileBase::time_point_t lastWriteTime = {};
+				};
+
+				struct SMacroData {
+					std::string identifier;
+					std::string definition;
+				};
+
+				struct SPreprocessorData {
+
+					inline bool operator==(const SPreprocessorData& other) const {
+						if (sourceIdentifier != other.sourceIdentifier) return false;
+						if (extraDefines.size() != other.extraDefines.size()) return false;
+						for (auto definesIt = extraDefines.begin(), otherDefinesIt = other.extraDefines.begin(); definesIt != extraDefines.end(); definesIt++, otherDefinesIt++) {
+							if (definesIt->identifier != otherDefinesIt->identifier || definesIt->definition != otherDefinesIt->definition) return false;
+						}
+						return true;
+					}
+
+					// We don't need to hash the preprocessor on its own but in conjunction with the compiler options. We return data to perform the hash on from SCompilerOptions
+					inline std::vector<char> getHashable() const
+					{
+						std::vector<char> hashable;
+						hashable.insert(hashable.end(), sourceIdentifier.data()[0], sourceIdentifier.data()[sourceIdentifier.size()]);
+						core::vector<SMacroData> sortedExtraDefines;
+						sortedExtraDefines.assign(extraDefines.begin(), extraDefines.end());
+						// Sort them by identifier so the hash is not order-sensitive!
+						std::sort(sortedExtraDefines.begin(), sortedExtraDefines.end(), [](const SMacroData& lhs, const SMacroData& rhs) {return lhs.identifier < rhs.identifier; });
+						for (const auto& defines : sortedExtraDefines) {
+							hashable.insert(hashable.end(), defines.identifier.data()[0], defines.identifier.data()[defines.identifier.size()]);
+							hashable.insert(hashable.end(), defines.definition.data()[0], defines.definition.data()[defines.definition.size()]);
+						}
+
+						return hashable;
+					}
```

Comment:
make this struct final, this method private and `SCompilerData` a friend

Reply:
tbf if the hasing gets done in `SCompilerData` ctor only, then you don't even need the method

Reply:
Move all this code to the `SCompilerData` hash method then?